File: apps/android/app/src/test/kotlin/com/loonext/android/core/contacts/VCardPropertiesTest.kt
package com.loonext.android.core.contacts

import java.io.File
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * #248 ROUND 3 — the vCard door's declaration, and the port of the property
 * enumeration it rests on.
 *
 * This door had no gate of any kind. `CATEGORIES:DNC`, `NOTE:DO NOT CONTACT -
 * asked us to stop`, and a label like `X-ABLabel=DO NOT CALL` on the TEL line
 * are where a .vcf says do-not-text; they are what Apple and Google actually
 * export, and all three were dropped without a word while the file's consent
 * attestation was written over the top.
 *
 * THE ENUMERATION HAS TO BE EXACT, in both directions, and neither failure is
 * quiet. A property this app MISSES is a refusal nobody can answer from the
 * phone — the server names a property the sheet never showed. A property this
 * app INVENTS is a question about something that is not in their file. So the
 * parity tests read `apps/api/src/routes/core/vcard.ts` and the shared contract,
 * and the behaviour tests below run the same shapes that parser handles.
 */
class VCardPropertiesTest {

    private fun repoFile(relative: String): String {
        var dir: File? = File("").absoluteFile
        while (dir != null) {
            val candidate = File(dir, relative)
            if (candidate.exists()) return candidate.readText()
            dir = dir.parentFile
        }
        throw AssertionError("$relative not found walking up from ${File("").absolutePath}")
    }

    private fun contractSource(): String = repoFile("packages/shared/src/contact-import.ts")

    private fun stringConst(source: String, name: String): String =
        Regex("""const $name\s*=\s*\n?\s*"([^"]*)"""").find(source)?.groupValues?.get(1)
            ?: throw AssertionError("$name is not a string constant in the shared module")

    // ------------------------------------------------------------- parity

    @Test
    fun `the properties the importer reads match the shared module`() {
        // Everything NOT on this list has to be declared. A phone holding a
        // longer list than the server would skip asking about a property the
        // server then refuses the file over, with no control anywhere to answer
        // it — the file becomes unimportable from this device.
        val block = Regex("""VCARD_MAPPED_PROPERTIES: readonly string\[\] = \[([^\]]*)\]""")
            .find(contractSource())
            ?.groupValues?.get(1)
            ?: throw AssertionError("VCARD_MAPPED_PROPERTIES is not a list literal")
        val shared = Regex("\"([^\"]*)\"").findAll(block).map { it.groupValues[1] }.toList()
        assertEquals(shared, VCardProperties.MAPPED)
    }

    @Test
    fun `both answers match the shared module`() {
        val body = Regex("""const VCARD_ACTIONS = new Set<string>\(\[([^\]]*)\]\)""")
            .find(contractSource())
            ?.groupValues?.get(1)
            ?: throw AssertionError("VCARD_ACTIONS is not a Set literal")
        val shared = body
            .split(",")
            .map { it.trim() }
            .filter { it != "" }
            .map { if (it.startsWith("\"")) it.trim('"') else stringConst(contractSource(), it) }
            .toSet()
        assertEquals(shared, VCardProperties.ACTIONS.toSet())
        // TWO answers, not the CSV's eight. A vCard property is present on a card
        // or it is not — there is no field to route it into — so the answers that
        // mean anything are "says nothing" and "do not text these".
        assertEquals(2, VCardProperties.ACTIONS.size)
    }

    @Test
    fun `the wire form is the one the server parses`() {
        val template = Regex("""return `([^`]+)`;""")
            .find(contractSource().substringAfter("export function formatVCardProperty"))
            ?.groupValues?.get(1)
            ?: throw AssertionError("formatVCardProperty does not return a template")
        val expected = template
            .replace("\${declaration.property}", "CATEGORIES")
            .replace("\${declaration.action}", VCardProperties.ACTION_OPTED_OUT)
        assertEquals(
            expected,
            VCardProperties.format("CATEGORIES", VCardProperties.ACTION_OPTED_OUT),
        )
        assertFalse(expected, expected.contains("\${"))
    }

    @Test
    fun `the parameter token is spelled the way the server spells it`() {
        // Built by SUBSTITUTING into the shared template rather than by retyping
        // `$property;$parameter` here. These two strings have to match character
        // for character in a place nothing else checks: a phone asking about
        // `TEL/TYPE` while the server waits to hear about `TEL;TYPE` is a client
        // whose every .vcf import is refused, with the person having answered
        // every question the sheet put in front of them.
        val template = Regex("""return `([^`]+)`;""")
            .find(contractSource().substringAfter("export function vcardParameterProperty"))
            ?.groupValues?.get(1)
            ?: throw AssertionError("vcardParameterProperty does not return a template")
        val expected = template
            .replace("\${property}", "TEL")
            .replace("\${parameter}", "X-ABLABEL")
        assertEquals(expected, VCardProperties.parameterProperty("TEL", "X-ABLABEL"))
        assertFalse(expected, expected.contains("\${"))
        // And the wire form still round-trips one: the declaration splits on the
        // LAST colon, so a token carrying a `;` is unremarkable to it.
        assertEquals(
            "TEL;X-ABLABEL" to VCardProperties.ACTION_OPTED_OUT,
            VCardProperties.parse(
                VCardProperties.format(
                    VCardProperties.parameterProperty("TEL", "X-ABLABEL"),
                    VCardProperties.ACTION_OPTED_OUT,
                ),
            ),
        )
    }

    // --------------------------------------------------- the enumeration

    /**
     * One file exercising every branch of the walk at once (rule 3): a card with
     * properties we read and properties we do not, a grouped property, a
     * parameterised one, a folded value whose continuation contains a colon, a
     * second card carrying something the first did not, and a line outside any
     * card at all.
     */
    private val mixedCards =
        "X-EVOLUTION-JUNK:outside any card\r\n" +
            "BEGIN:VCARD\r\n" +
            "VERSION:3.0\r\n" +
            "FN:Ann Fischer\r\n" +
            "TEL;TYPE=CELL:+14165550100\r\n" +
            "CATEGORIES:DNC\r\n" +
            "NOTE:DO NOT CONTACT - asked us\r\n" +
            "  to stop: 2026-01-04\r\n" +
            "item1.X-ABLabel:Home\r\n" +
            "END:VCARD\r\n" +
            "BEGIN:VCARD\r\n" +
            "VERSION:3.0\r\n" +
            "FN:Bob Vance\r\n" +
            "TEL:+14165550101\r\n" +
            "BDAY:1980-04-01\r\n" +
            "END:VCARD\r\n"

    @Test
    fun `every property the cards carry that we do not read is reported`() {
        assertEquals(
            listOf("BDAY", "CATEGORIES", "NOTE", "TEL;TYPE", "X-ABLABEL"),
            VCardProperties.undeclared(mixedCards.toByteArray()),
        )
    }

    @Test
    fun `the two properties that can say do-not-text are both in the list`() {
        // The whole reason this door has a gate now. Named individually rather
        // than left to the list above, because the list changing for an unrelated
        // reason must not quietly take one of these two with it.
        val found = VCardProperties.undeclared(mixedCards.toByteArray())
        assertTrue("CATEGORIES:DNC must reach a person", found.contains("CATEGORIES"))
        assertTrue("a NOTE saying they asked us to stop must too", found.contains("NOTE"))
    }

    @Test
    fun `a folded line is not read as a second property`() {
        // THE trap in this port. The continuation `  to stop: 2026-01-04` carries
        // a colon, so a walk that did not unfold would invent a property called
        // "TO STOP" — a question about something that is not in their file — and
        // the real NOTE would still be there underneath it.
        val found = VCardProperties.undeclared(mixedCards.toByteArray())
        assertFalse(found.toString(), found.any { it.contains("TO STOP") })
        // And the unfolding must not swallow the property either: a file whose
        // only interesting line is folded still asks about it.
        assertEquals(
            listOf("NOTE"),
            VCardProperties.undeclared(
                (
                    "BEGIN:VCARD\nFN:Ann\nTEL:+14165550100\n" +
                        "NOTE:they asked us\n to stop texting\nEND:VCARD\n"
                    ).toByteArray(),
            ),
        )
    }

    @Test
    fun `a grouped or parameterised property is named the way the server names it`() {
        // `item1.X-ABLabel` is what Apple exports and the server strips the group
        // prefix and upper-cases it. A phone that sent `item1.X-ABLabel:ignore`
        // would be answering about a property the server does not have, and the
        // file would be refused with the person having answered correctly.
        assertTrue(
            VCardProperties.undeclared(mixedCards.toByteArray()).contains("X-ABLABEL"),
        )
        assertEquals(
            listOf("X-SOCIALPROFILE", "X-SOCIALPROFILE;TYPE"),
            VCardProperties.undeclared(
                (
                    "BEGIN:VCARD\nFN:Ann\nTEL:+1\n" +
                        "item2.X-socialProfile;TYPE=twitter:https://x.com/ann\nEND:VCARD\n"
                    ).toByteArray(),
            ),
        )
    }

    @Test
    fun `a parameter is its own question, qualified by the property it sits on`() {
        // APPLE'S INLINE SHAPE, and the reason parameters are enumerated at all.
        // The property is TEL, TEL is read, and everything after the first `;`
        // used to be thrown away — so the one sentence on the line saying not to
        // text this person was the one part nobody looked at. The grouped
        // `item1.X-ABLabel:` form was always caught, which is exactly what made
        // this one look covered.
        assertEquals(
            listOf("TEL;TYPE", "TEL;X-ABLABEL"),
            VCardProperties.undeclared(
                (
                    "BEGIN:VCARD\r\nFN:Ann Fischer\r\n" +
                        "TEL;TYPE=CELL;X-ABLabel=DO NOT CALL:+16135550100\r\nEND:VCARD\r\n"
                    ).toByteArray(),
            ),
        )
        // Qualified by the PROPERTY, never bare: a TYPE on TEL and a TYPE on
        // EMAIL are different text on different lines, and one answer covering
        // both would dismiss a value nobody saw.
        assertEquals(
            listOf("EMAIL", "EMAIL;TYPE", "TEL;TYPE"),
            VCardProperties.undeclared(
                (
                    "BEGIN:VCARD\r\nFN:Ann\r\nTEL;TYPE=CELL:+1416\r\n" +
                        "EMAIL;TYPE=WORK:ann@example.com\r\nEND:VCARD\r\n"
                    ).toByteArray(),
            ),
        )
        // A MIX down one line: a valueless parameter (`PREF` is its own name,
        // there is no `=` to cut at), one that has a value, and a repeat that
        // must not be asked about twice. Three different branches of the segment
        // loop, on the line shape a real export writes.
        assertEquals(
            listOf("TEL;PREF", "TEL;TYPE"),
            VCardProperties.undeclared(
                (
                    "BEGIN:VCARD\r\nFN:Ann\r\n" +
                        "TEL;PREF;TYPE=CELL;TYPE=VOICE:+1416\r\nEND:VCARD\r\n"
                    ).toByteArray(),
            ),
        )
    }

    @Test
    fun `a line the format calls nonsense is still a line somebody wrote`() {
        // Two doors, both of which delivered a message. A line with NO colon is
        // not a content line by the RFC — a statement about the format, not about
        // what the file was trying to say. And a malformed parameter leaves no
        // unquoted colon either, so `CATEGORIES` (one of the two places a .vcf can
        // say stop) went unasked because of a typo.
        assertEquals(
            listOf("DO-NOT-CALL"),
            VCardProperties.undeclared(
                "BEGIN:VCARD\r\nFN:Ann\r\nTEL:+1416\r\nDO-NOT-CALL\r\nEND:VCARD\r\n".toByteArray(),
            ),
        )
        assertEquals(
            listOf("CATEGORIES", "CATEGORIES;TYPE"),
            VCardProperties.undeclared(
                (
                    "BEGIN:VCARD\r\nFN:Ann\r\nTEL:+1416\r\n" +
                        "CATEGORIES;TYPE=\"a:DNC\r\nEND:VCARD\r\n"
                    ).toByteArray(),
            ),
        )
    }

    @Test
    fun `a plain export asks about its parameters and nothing else`() {
        // THE ORDINARY CASE, AND IT IS NO LONGER SILENT. This test used to assert
        // that an FN/N/TEL export asked nothing, over a fixture whose TEL line
        // carries `TYPE=CELL` — which stopped being true the day a parameter
        // became something to declare, and would have held this port at last
        // month's answer while every .vcf import from this phone was refused.
        //
        // The cost is stated rather than exempted: a parameter is free text and
        // `TYPE=DNC` is a real export, so a rule letting the ubiquitous ones
        // through would be a vocabulary.
        assertEquals(
            listOf("TEL;TYPE"),
            VCardProperties.undeclared(
                (
                    "BEGIN:VCARD\r\nVERSION:3.0\r\nN:Fischer;Ann;;;\r\nFN:Ann Fischer\r\n" +
                        "TEL;TYPE=CELL:+14165550100\r\nEND:VCARD\r\n"
                    ).toByteArray(),
            ),
        )
        // A card with no parameters on it still asks nothing, so the sheet only
        // appears when the file gives it something to say.
        assertEquals(
            emptyList<String>(),
            VCardProperties.undeclared(
                (
                    "BEGIN:VCARD\r\nVERSION:3.0\r\nN:Fischer;Ann;;;\r\nFN:Ann Fischer\r\n" +
                        "TEL:+14165550100\r\nEND:VCARD\r\n"
                    ).toByteArray(),
            ),
        )
    }

    @Test
    fun `nothing outside a card is a property of one`() {
        // Only content lines INSIDE a BEGIN/END block count, exactly as the
        // server counts them — noise around the cards is not a question.
        assertFalse(
            VCardProperties.undeclared(mixedCards.toByteArray()).contains("X-EVOLUTION-JUNK"),
        )
    }

    // -------------------------------------------------------- format/parse

    @Test
    fun `a declaration round-trips, splitting on the last colon`() {
        val wire = VCardProperties.format("CATEGORIES", VCardProperties.ACTION_OPTED_OUT)
        assertEquals("CATEGORIES" to VCardProperties.ACTION_OPTED_OUT, VCardProperties.parse(wire))
        // The LAST colon, the mirror of the column format's first: a property name
        // may be grouped or parameterised, and the action is the fixed token, so
        // the fixed end is the safe end to split from.
        assertEquals(
            "ITEM1.X-ABLABEL" to VCardProperties.ACTION_IGNORE,
            VCardProperties.parse("item1.X-ABLabel:${VCardProperties.ACTION_IGNORE}"),
        )
        assertNull(VCardProperties.parse("CATEGORIES"))
        assertNull("an action nobody knows is not an answer", VCardProperties.parse("NOTE:maybe"))
        assertNull(VCardProperties.parse(":ignore"))
        // The CSV's field names are not answers here, and must not be smuggled in:
        // a vCard property routed to `notes` would be read as nothing.
        assertNull(VCardProperties.parse("NOTE:notes"))
    }
}

File: apps/web/src/lib/contacts/vcard-properties.test.ts
import {
  CONTACT_IMPORT_IGNORE,
  formatVCardProperty,
  vcardParameterProperty,
  VCARD_MAPPED_PROPERTIES,
} from "@loonext/shared";
import { describe, expect, it } from "vitest";

import {
  answerProperty,
  ignoreRemainingProperties,
  propertyDeclarations,
  readVCardProperties,
  VCARD_SAMPLE_LIMIT,
} from "./vcard-properties";

/**
 * #248 round 3 — the vCard door, which had no gate of any kind.
 *
 * `CATEGORIES:DNC` and a `NOTE` reading "DO NOT CONTACT - asked us to stop" are
 * the only two places the format lets a card say do-not-text, they are what
 * Apple and Google actually export, and both were dropped by the importer
 * without a word while the file's consent attestation was written over the top.
 *
 * The fixture below is a MIX: three cards, one carrying a category, one
 * carrying a note, one carrying neither. A one-card fixture is how the CSV half
 * of this issue passed twice while broken.
 */
const BOOK = [
  "BEGIN:VCARD",
  "VERSION:3.0",
  "FN:Ann Rivera",
  "TEL;TYPE=CELL:+12065550100",
  "CATEGORIES:DNC",
  "ORG:Rivera Roofing",
  "END:VCARD",
  "BEGIN:VCARD",
  "VERSION:3.0",
  "FN:Bo Tran",
  "TEL;TYPE=CELL:+12065550101",
  "NOTE:DO NOT CONTACT - asked us to stop",
  "END:VCARD",
  "BEGIN:VCARD",
  "VERSION:3.0",
  "FN:Cy Okafor",
  "TEL;TYPE=CELL:+12065550102",
  "ORG:Okafor Electric",
  "END:VCARD",
].join("\r\n");

describe("readVCardProperties: what the cards carry that we do not read", () => {
  it("VP-1: reports the two properties a .vcf can say do-not-text in", () => {
    const { cards, properties } = readVCardProperties(BOOK);

    expect(cards).toBe(3);
    // `TEL;TYPE` is in this list because a PARAMETER is free text of its own.
    // TEL is mapped, so its line used to be taken for the number and the rest
    // of it thrown away, which is where Apple writes `X-ABLabel=DO NOT CALL`.
    // The cost of having no vocabulary is that the ubiquitous `TYPE=CELL` gets
    // asked about too; that cost is accepted upstream, in the shared docblock.
    expect(properties.map((row) => row.property).sort()).toEqual([
      "CATEGORIES",
      "NOTE",
      "ORG",
      vcardParameterProperty("TEL", "TYPE"),
    ]);
  });

  it("VP-2: shows what each one SAYS, sentence and all", () => {
    // The values are the whole design. A person who cannot see "DNC" cannot
    // skip it knowingly, and then the click is theatre. Nothing truncates them
    // either: round two capped a decision at 16 characters, and this note is
    // 36 — "a decision is a token, not a sentence" is simply false.
    const { properties } = readVCardProperties(BOOK);
    const byName = new Map(properties.map((row) => [row.property, row]));

    expect(byName.get("CATEGORIES")?.samples).toEqual(["DNC"]);
    expect(byName.get("NOTE")?.samples).toEqual([
      "DO NOT CONTACT - asked us to stop",
    ]);
  });

  it("VP-3: never asks about a property the importer reads", () => {
    // Asserted against the SHARED list the server excludes by, not a list
    // retyped here: a property the importer starts reading tomorrow must stop
    // being asked about in the same commit, and a retyped copy would go on
    // asking.
    const asked = readVCardProperties(BOOK).properties.map(
      (row) => row.property,
    );

    for (const mapped of VCARD_MAPPED_PROPERTIES) {
      expect(asked).not.toContain(mapped);
    }
    // And the envelope itself is not a fact about a person.
    expect(asked).not.toContain("VERSION");
  });

  it("VP-4: pre-answers nothing at all", () => {
    // Unlike the CSV wizard, where a header spelled `Phone` is a guess worth
    // making on somebody's behalf, there is nothing to guess here: `NOTE` is
    // house-keeping on most cards and a revocation on the one that matters, and
    // this product may not pick either answer for a customer.
    const { properties } = readVCardProperties(BOOK);

    expect(properties.every((row) => row.answer === null)).toBe(true);
    expect(propertyDeclarations(properties)).toBeNull();
  });

  it("VP-5: counts the cards each property is actually on", () => {
    // "On 1 of 40 cards" is the difference between the file's furniture and
    // somebody having said something about one person.
    const byName = new Map(
      readVCardProperties(BOOK).properties.map((row) => [row.property, row]),
    );

    expect(byName.get("ORG")?.cards).toBe(2);
    expect(byName.get("CATEGORIES")?.cards).toBe(1);
    expect(byName.get("NOTE")?.cards).toBe(1);
    expect(byName.get(vcardParameterProperty("TEL", "TYPE"))?.cards).toBe(3);
    // RAREST FIRST. `CATEGORIES:DNC` is on one card of three and is the reason
    // this screen exists; `TEL;TYPE` is on all three and decides nothing. The
    // order used to be most-carried first, which was survivable while `ORG` led
    // it and became the file's furniture burying the one-off the moment
    // parameters were enumerated.
    const order = readVCardProperties(BOOK).properties.map(
      (row) => row.property,
    );
    expect(order[0]).toBe("CATEGORIES");
    expect(order[order.length - 1]).toBe(vcardParameterProperty("TEL", "TYPE"));
  });

  it("VP-6: reads a grouped, parameterised property under its bare name", () => {
    // Apple exports `item1.X-ABLabel:` and `item1.X-ABADR;TYPE=HOME:`. Reading
    // those as `ITEM1.X-ABLABEL` would declare a property name the server never
    // reports, so its own would go unanswered and every Apple export would be
    // refused with nothing the person could do about it.
    const { properties } = readVCardProperties(
      [
        "BEGIN:VCARD",
        "VERSION:3.0",
        "FN:Ann",
        "TEL:+12065550100",
        "item1.X-ABLabel:do not text",
        "END:VCARD",
      ].join("\r\n"),
    );

    expect(properties.map((row) => row.property)).toEqual(["X-ABLABEL"]);
    expect(properties[0].samples).toEqual(["do not text"]);
  });

  it("VP-7: unfolds a wrapped line instead of reading it as a property", () => {
    // RFC folding breaks a long NOTE across lines with a leading space. Read
    // unfolded, the tail is part of the note; read raw, the note is truncated
    // exactly where the do-not-text part begins.
    const { properties } = readVCardProperties(
      [
        "BEGIN:VCARD",
        "VERSION:3.0",
        "FN:Ann",
        "TEL:+12065550100",
        "NOTE:Roof done May 2026. Asked us",
        "  to stop texting her.",
        "END:VCARD",
      ].join("\r\n"),
    );

    expect(properties).toHaveLength(1);
    expect(properties[0].samples).toEqual([
      "Roof done May 2026. Asked us to stop texting her.",
    ]);
  });

  it("VP-8: reports distinct values and says when there are more", () => {
    const { properties } = readVCardProperties(
      [
        "BEGIN:VCARD",
        "FN:A",
        "TEL:+12065550100",
        "X-STATUS:one",
        "END:VCARD",
        "BEGIN:VCARD",
        "FN:B",
        "TEL:+12065550101",
        "X-STATUS:two",
        "END:VCARD",
        "BEGIN:VCARD",
        "FN:C",
        "TEL:+12065550102",
        "X-STATUS:three",
        "END:VCARD",
        "BEGIN:VCARD",
        "FN:D",
        "TEL:+12065550103",
        "X-STATUS:DO NOT CALL",
        "END:VCARD",
      ].join("\r\n"),
    );

    expect(properties[0].samples).toHaveLength(VCARD_SAMPLE_LIMIT);
    expect(properties[0].more).toBe(true);
    expect(properties[0].cards).toBe(4);
  });

  it("VP-14: reads the LABEL on a TEL line, not the number beside it", () => {
    // The third door, and the one that looked closed. `item1.X-ABLabel:` — the
    // grouped form — was always caught; this is Apple's OTHER shape, inline on
    // the TEL line. TEL is mapped, so the line was read for its number and
    // everything after the first `;` was discarded, which is exactly where the
    // one sentence saying not to text this person was sitting.
    const { properties } = readVCardProperties(
      [
        "BEGIN:VCARD",
        "VERSION:3.0",
        "FN:Ann",
        "TEL;TYPE=CELL;X-ABLabel=DO NOT CALL:+16135550100",
        "END:VCARD",
        "BEGIN:VCARD",
        "VERSION:3.0",
        "FN:Bo",
        "TEL;TYPE=HOME:+16135550101",
        "END:VCARD",
      ].join("\r\n"),
    );
    const byName = new Map(properties.map((row) => [row.property, row]));

    // The token the SERVER will demand, built by the shared function rather
    // than spelled here: a client that declares `TEL;X-ABLabel=DO NOT CALL` has
    // answered a question nobody asked, and the upload is refused with nothing
    // the person can do about it.
    const label = vcardParameterProperty("TEL", "X-ABLABEL");
    expect(byName.get(label)?.samples).toEqual(["DO NOT CALL"]);
    // The parameter's own text, never the line's value. A phone number is not
    // evidence about the label sitting beside it.
    expect(byName.get(label)?.samples).not.toContain("+16135550100");
    expect(byName.get(label)?.cards).toBe(1);
    // The mix: the ordinary card carries TYPE and no label, so the two
    // parameters on one line are counted apart.
    expect(byName.get(vcardParameterProperty("TEL", "TYPE"))?.samples).toEqual([
      "CELL",
      "HOME",
    ]);
    // And the one-off leads, not the parameter every card carries.
    expect(properties[0].property).toBe(label);
  });

  it("VP-15: enumerates a line with no colon at all", () => {
    // `DO-NOT-CALL` on its own is not a content line by the RFC, which is a
    // statement about the format rather than about what the file was trying to
    // say. The old parser returned null before the property was ever recorded,
    // so the card imported and the message was delivered.
    const { properties } = readVCardProperties(
      [
        "BEGIN:VCARD",
        "VERSION:3.0",
        "FN:Ann",
        "TEL:+12065550100",
        "DO-NOT-CALL",
        "END:VCARD",
      ].join("\r\n"),
    );

    expect(properties.map((row) => row.property)).toEqual(["DO-NOT-CALL"]);
    // Nothing to show, only something to declare: the token IS the whole line.
    expect(properties[0].samples).toEqual([]);
    expect(properties[0].cards).toBe(1);
    expect(propertyDeclarations(properties)).toBeNull();
  });

  it("VP-16: enumerates a property whose parameter is malformed", () => {
    // `CATEGORIES;TYPE="a:DNC` — an unbalanced quote, so the only colon on the
    // line reads as quoted and there is no value. Same drop by a different
    // route, and the property is one of the two a .vcf can say stop in.
    const { properties } = readVCardProperties(
      [
        "BEGIN:VCARD",
        "VERSION:3.0",
        "FN:Ann",
        "TEL:+12065550100",
        'CATEGORIES;TYPE="a:DNC',
        "END:VCARD",
      ].join("\r\n"),
    );

    expect(properties.map((row) => row.property).sort()).toEqual([
      "CATEGORIES",
      vcardParameterProperty("CATEGORIES", "TYPE"),
    ]);
  });

  it("VP-9: tolerates a file with no VCARD blocks at all", () => {
    // The server refuses this file separately ("no VCARD blocks found"). This
    // one must not invent properties out of the loose lines on the way there.
    expect(readVCardProperties("FN:Ann\r\nTEL:+12065550100")).toEqual({
      cards: 0,
      properties: [],
    });
  });
});

describe("answering the properties", () => {
  it("VP-10: declares every property once the last one is answered", () => {
    let properties = readVCardProperties(BOOK).properties;
    for (const row of properties) {
      properties = answerProperty(
        properties,
        row.property,
        row.property === "CATEGORIES" ? "opted_out" : CONTACT_IMPORT_IGNORE,
      );
    }

    expect(propertyDeclarations(properties)?.map(formatVCardProperty)).toEqual([
      "CATEGORIES:opted_out",
      `NOTE:${CONTACT_IMPORT_IGNORE}`,
      `ORG:${CONTACT_IMPORT_IGNORE}`,
      `${vcardParameterProperty("TEL", "TYPE")}:${CONTACT_IMPORT_IGNORE}`,
    ]);
  });

  it("VP-11: one answer short is no declaration at all", () => {
    // Null rather than a partial list, because a partial list is exactly what
    // the server refuses — and a client that posts one is betting the refusal
    // will be read by somebody.
    const properties = answerProperty(
      readVCardProperties(BOOK).properties,
      "ORG",
      CONTACT_IMPORT_IGNORE,
    );

    expect(propertyDeclarations(properties)).toBeNull();
  });

  it("VP-12: the bulk dismissal never overwrites an answer somebody gave", () => {
    const answered = answerProperty(
      readVCardProperties(BOOK).properties,
      "CATEGORIES",
      "opted_out",
    );
    const rest = ignoreRemainingProperties(answered);

    expect(rest.find((row) => row.property === "CATEGORIES")?.answer).toBe(
      "opted_out",
    );
    expect(propertyDeclarations(rest)?.map(formatVCardProperty)).toContain(
      "CATEGORIES:opted_out",
    );
  });

  it("VP-13: an answer can be taken back", () => {
    // The row stays listed after it is answered. One that vanished would leave
    // no way to see what was asserted, or to change your mind.
    const answered = ignoreRemainingProperties(
      readVCardProperties(BOOK).properties,
    );
    const undone = answerProperty(answered, "NOTE", null);

    expect(propertyDeclarations(undone)).toBeNull();
  });
});
